tools/ci_build/rewriter.py

```diff
@@ -0,0 +1,226 @@
+                    for subgraph in attr.graphs:
+                        extract_ops_from_graph(subgraph, opsets, operators)
+
+    for root, _, files in os.walk(model_path):
+        for file in files:
+            if file.endswith('.onnx'):
+                model_path = os.path.join(root, file)
+                model = onnx.load(model_path)
+                extract_ops_from_graph(model.graph, model.opset_import, referred_ops)
+
+    return referred_ops #end of extract_ops_from_model(...)
+
+
+def rewrite_provider(model_path, file_path, ep_path):
+    '''rewrite provider file to exclude unused ops'''
+
+    onnx_op = 'ONNX_OPERATOR_KERNEL_CLASS_NAME'
+    onnx_op_len = len(onnx_op)
+    onnx_typed_op = 'ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME'
+    onnx_typed_op_len = len(onnx_typed_op)
+    onnx_versioned_op = 'ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME'
+    onnx_versioned_op_len = len(onnx_versioned_op)
+    onnx_versioned_typed_op = 'ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME'
+    onnx_versioned_typed_op_len = len(onnx_versioned_typed_op)
+    version_map = {} #{op:{domain:[opset1, opset2, opset3 ...]}
+    operators = extract_ops_from_file(file_path, extract_ops_from_model(model_path, {}))
+
+    def fill_version_map(op_type, opset_from, opset_to, domain):
+        '''callback func to register op in version_map'''
+
+        if op_type in version_map:
+            if domain in version_map[op_type]:
+
+                if opset_from not in version_map[op_type][domain]:
+                    version_map[op_type][domain].append(opset_from)
+
+                if opset_to not in version_map[op_type][domain]:
+                    version_map[op_type][domain].append(opset_to)
+            else:
+                version_map[op_type][domain] =\
+                    [opset_from, opset_to] if opset_from < opset_to else [opset_from]
+
+        else:
+            version_map[op_type] =\
+                {domain: [opset_from, opset_to] if opset_from < opset_to else [opset_from]}
+
+        return True #end of fill_version_map(...)
+
+
+    def need_comment(op_type, opset_from, opset_to, domain):
+        '''callback func to check if the op is in ops'''
+
+        if op_type in operators and domain in operators[op_type]:
+            opset_to_index = version_map[op_type][domain].index(opset_to)
+
+            if opset_to_index == len(version_map[op_type][domain]) - 1:
+                opset_to = 9999 #if is the latest, extend to unlimited
+
+            elif opset_from == opset_to:
+                opset_to = version_map[op_type][domain][opset_to_index + 1]
+
+            for opset in range(opset_from, opset_to + 1):
+                if opset in operators[op_type][domain]:
+                    return False #do not comment
+
+        return True #end of need_comment(...)
+
+
+    def process_lines(lines, offset, end_mark, call_back):
+        '''extract op info from a logic code line start from offset to the line end
```

Comment:
Having a comment that shows the text of the lines you're parsing would be very helpful to someone reading the script (vs. making them go fine the C++ file and the correct places in it).

Reply:
Well the cc file changes constantly - that case very likely user find wrong lines if not synced timely.
Guess the comment good enough no?

Reply:
I was thinking something like the below, so the reader knows what you're intending to process, and if the source file changes in some unexpected way it's easy to see the diff vs what this script is expecting. There's quite a bit of detail to the parsing that isn't obvious as the forward declaration and BuildKernelCreateInfo lines have 4 variants - typed and untyped, old version with range and current version. It makes it hard for the next developer who has to work on this script to understand what is being parsed for in order to make any necessary changes.

e.g. 

Current version, un-typed and typed
# class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 1, Transpose);
# class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 8, float, Expand);
Old version
# class ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 1, 10, Unsqueeze);
# class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 7, 9, float, Upsample);

Similar for BuildKernelCreateInfo
# BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 7, Cos)>,
# BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 1, 10, Hardmax)>,
# BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 7, double, Sin)>,
# BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 1, 10, float, LogSoftmax)>,


---
In reply to: [468928027](https://github.com/microsoft/onnxruntime/pull/4747#discussion_r468928027) [](ancestors = 468928027)

Reply:
Cool, will add them as help.